File: src/main/scala/com/michalplachta/galactic/console/GalacticFollowersApp.scala
package com.michalplachta.galactic.console

import com.michalplachta.galactic.service.FollowersService
import com.michalplachta.galactic.service.FollowersService.Version4._

import scala.annotation.tailrec
import scala.util.{ Failure, Success }

object GalacticFollowersApp extends App {
  @tailrec
  def runFollowersApp(getFollowersText: String ⇒ String): Unit = {
    println("---\nEnter Citizen's name: ")
    val citizenName = io.StdIn.readLine()
    println(s"Getting followers for $citizenName")
    val followersDescription = getFollowersText(citizenName)
    println(s"$citizenName has $followersDescription followers!")
    runFollowersApp(getFollowersText)
  }

  def getAndDescribe(citizenName: String): String = {
    FollowersService.Version1.getFollowers(citizenName).toString
  }

  def getAndDescribeUsingCache(citizenName: String): String = {
    val cachedFollowers = FollowersService.Version2.getCachedFollowers(citizenName)
    cachedFollowers.map(_.toString).getOrElse("(not available)")
  }

  def getAndDescribeUsingCacheWithFailures(citizenName: String): String = {
    val cachedTriedFollowers = FollowersService.Version3.getCachedTriedFollowers(citizenName)
    cachedTriedFollowers map {
      case Success(followers)    ⇒ followers.toString
      case Failure(errorMessage) ⇒ s"(failed to get followers: $errorMessage)"
    } getOrElse "(not available)"
  }

  def getAndDescribeUsingADTs(citizenName: String): String = {
    val remoteFollowers: RemoteFollowersData = FollowersService.Version4.getRemoteFollowers(citizenName)
    remoteFollowers match {
      case NotRequestedYet()    ⇒ "(not requested yet)"
      case Loading()            ⇒ "(loading...)"
      case Fetched(followers)   ⇒ followers.toString
      case Failed(errorMessage) ⇒ s"(failed to get followers: $errorMessage)"
    }
  }

  runFollowersApp(getAndDescribeUsingADTs)
}

File: src/main/java/com/michalplachta/galactic/java/console/GalacticFollowersApp.java
package com.michalplachta.galactic.java.console;

import com.michalplachta.galactic.java.service.FollowersService;
import com.michalplachta.galactic.java.service.FollowersService.Version4.RemoteFollowersData;
import io.vavr.control.Option;
import io.vavr.control.Try;

import java.util.Scanner;
import java.util.function.Function;

import static com.michalplachta.galactic.java.internal.RemoteFollowersDataPatterns.*;
import static io.vavr.API.*;
import static io.vavr.Patterns.$Failure;
import static io.vavr.Patterns.$Success;

public class GalacticFollowersApp {
    private static void runFollowersApp(Function<String, String> getFollowersText) {
        System.out.println("---\nEnter Citizen's name: ");
        Scanner scanner = new Scanner(System.in);
        String citizenName = scanner.nextLine();
        System.out.println(String.format("Getting followers for %s", citizenName));
        String followersText = getFollowersText.apply(citizenName);
        System.out.println(String.format("%s has %s followers!", citizenName, followersText));
        runFollowersApp(getFollowersText);
    }

    private static String getAndDescribe(String citizenName) {
        return FollowersService.Version1.getFollowers(citizenName).toString();
    }

    private static String getAndDescribeUsingCache(String citizenName) {
        Option<Integer> cachedFollowers = FollowersService.Version2.getCachedFollowers(citizenName);
        return cachedFollowers.map(Object::toString).getOrElse("(loading...)");
    }

    private static String getAndDescribeUsingCacheWithFailures(String citizenName) {
        Option<Try<Integer>> cachedTriedFollowers = FollowersService.Version3.getCachedTriedFollowers(citizenName);
        return cachedTriedFollowers.map(triedFollowers -> Match(triedFollowers).of(
            Case($Success($()), Object::toString),
            Case($Failure($()), errorMessage -> String.format("(failed to get followers: %s)", errorMessage))
        )).getOrElse("(loading...)");
    }

    private static String getAndDescribeUsingADTs(String citizenName) {
        RemoteFollowersData remoteFollowers = FollowersService.Version4.getRemoteFollowers(citizenName);
        return Match(remoteFollowers).of(
                Case(NotRequestedYet(), "(not requested yet)"),
                Case(Loading(), "(loading...)"),
                Case(Fetched($()), Object::toString),
                Case(Failed($()), errorMessage -> String.format("(failed to get followers: %s)", errorMessage))
        );
    }

    public static void main(String[] args){
        runFollowersApp(GalacticFollowersApp::getAndDescribeUsingADTs);
    }
}
